volatility3/framework/symbols/linux/utilities/module_extract.py

```diff
@@ -0,0 +1,891 @@
+                sym_type,
+                st_fmt,
+                module,
+            )
+            if not data:
+                vollog.debug(
+                    f"Coult not construct a symbol table for module at {module.vol.offset}. Cannot recover."
+                )
+                return None, None, None
+
+            symtab_index = len(updated_sections)
+
+            # Manually add symtab with the correct data
+            updated_sections.append((".symtab", symtab_address, file_offset, data))
+
+        else:
+            vollog.debug(
+                f"Did not find a .symtab section for module at {module.vol.offset:#x}. Cannot recover."
+            )
+            return None, None, None
+
+        return updated_sections, strtab_index, symtab_index
+
+    @classmethod
+    def _make_elf_header_32(
+        cls, sect_hdr_offset: int, num_sections: int
+    ) -> Optional[bytes]:
+        """
+        Creates a 32 bit ELF header for the file based on recovered values
+        Called last as it needs information computed from the sections
+
+        Spec: https://refspecs.linuxfoundation.org/elf/gabi4+/ch4.eheader.html
+        """
+        e_ident = b"\x7f\x45\x4c\x46\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+        e_type = b"\x01\x00"  # relocateble
+        e_machine = b"\x03\x00"  # EM_X86_86
+        e_version = b"\x01\x00\x00\x00"
+        e_entry = b"\x00" * 4  # The .init sections are freed after module load
+        e_phoff = b"\x00" * 4
+        e_shoff = struct.pack("<I", sect_hdr_offset)
+        e_flags = b"\x00\x00\x00\x00"
+        e_ehsize = b"\x34\x00"  # 52 - size of ELF Header
+        e_phentsize = b"\x00\x00"
+        e_phnum = b"\x00\x00"
+        e_shentsize = b"\x28\x00"
+        e_shnum = struct.pack("<H", num_sections + 1)
+        e_shstrndx = struct.pack("<H", num_sections)
+
+        header = (
+            e_ident
+            + e_type
+            + e_machine
+            + e_version
+            + e_entry
+            + e_phoff
+            + e_shoff
+            + e_flags
+            + e_ehsize
+            + e_phentsize
+            + e_phnum
+            + e_shentsize
+            + e_shnum
+            + e_shstrndx
+        )
+
+        # should never happen as we make the header ourselves
+        if len(header) != 52:
+            vollog.error(
+                f"_get_header_32 created a header of {len(header)} bytes. Cannot proceed"
+            )
+            return None
+
+        return header
+
+    @classmethod
+    def _make_elf_header_64(
+        cls, sect_hdr_offset: int, num_sections: int
+    ) -> Optional[bytes]:
+        """
+        Creates a 64 bit ELF header for the file based on recovered values
+        Called last as it needs information computed from the sections
+
+        Spec: https://refspecs.linuxfoundation.org/elf/gabi4+/ch4.eheader.html
+        """
+        e_ident = b"\x7f\x45\x4c\x46\x02\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00"
```

Comment:
Some of these would then (presumably) be shared with 32 bit if they were in a constants file?

Reply:
and these two could probably become a single method parameterized with a boolean to say 32/64?

Reply:
I normally like condensed code but it would be pretty ugly I think to split this. I could try though

Reply:
It wouldn't be splitting them up, it'd be a single function to do them both, but with a parameter that clearly shows which part of the structure changes based on 32 vs 64?  What I haven't checked is if there's extra fields, or anything weird that could make it harder to understand...

Reply:
ok I combined the making of section and elf header